WalletWasabi/ExchangeRate/ExchangeRateProvider.cs

```diff
@@ -0,0 +1,79 @@
+using System.Collections.Immutable;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+using WalletWasabi.Extensions;
+using ExchangeRateProviderInfo = (string Name, string ApiUrl, bool ClearNetOnly, System.Func<string, decimal> Extractor);
+
+namespace WalletWasabi.ExchangeRate;
+
+public class ExchangeRateProvider(EndPoint? socksProxyEndPoint = null)
+{
+	private static readonly string ProviderList = @"""
+	Name       ClearNetOnly	 ApiUrl	                                                                    JsonPath
+    ---------  ------------  -------------------------------------------------------------------------  ------------------
+	bitstamp   true          https://www.bitstamp.net/api/v2/ticker/btcusd                              .bid
+	blockchain false         https://blockchain.info/ticker                                             .USD.buy
+	coinbase   true          https://api.coinbase.com/v2/exchange-rates?currency=BTC                    .data.rates.USD
+	coingecko  false         https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&ids=bitcoin .[0].current_price
+	coingate   true          https://api.coingate.com/v2/rates/merchant/BTC/USD                         $
+	gemini     false         https://api.gemini.com/v1/pubticker/btcusd                                 .bid """;
+
+	public static readonly ImmutableArray<ExchangeRateProviderInfo> Providers =
+		ProviderList
+			.SplitLines()
+			.Skip(3)
+			.Select(l => l.SplitWords())
+			.Select(ws => new ExchangeRateProviderInfo(ws[0].Trim(), ws[2], bool.Parse(ws[1]), JsonPath(ws[3])))
+			.ToImmutableArray();
+
+	public async Task<ExchangeRate> GetExchangeRateAsync(string providerName, CancellationToken cancellationToken)
```

Comment:
Nit: you could use enum instead of providerName. Less error-prone and simplified code below with string comparison.

Reply:
I prefer the strings because I expect user be able to add other providers without touching the code.